docs/core/adding-spks.md

```diff
@@ -0,0 +1,683 @@
+---
+id: adding-spks
+title: Adding New Script Types
+---
+
+## Adding a New ScriptPubKey Type
+
+In this document, we will describe how to add new script implementations and types in Bitcoin-S. We will use the following script template example which we have called P2PK with Timeout to illustrate the process:
+
+```
+OP_IF
+    <Public Key>
+OP_ELSE
+    <Timeout> OP_CHECKSEQUENCEVERIFY OP_DROP
+    <Timeout Public Key>
+OP_ENDIF
+OP_CHECKSIG
+```
+
+Here is [the actual pull request](https://github.com/bitcoin-s/bitcoin-s/pull/967) in which a very similar `ScriptPubKey` is implemented in Bitcoin-S.
+
+Please note that this document only explains how to add new `RawScriptPubKey`s which are the subset of `ScriptPubKey`s which are fully described by their raw scripts. This is to say that this guide will not help in implementing a new segwit version, but should be helpful for most anything else.
+
+It is also important to note that all new scripts should be implemented as if they are to appear on-chain without any P2SH or P2WSH. Bitcoin-S already supports conversions from raw on-chain scripts to these formats in the constructors for the script hash schemes which does not require extra support for new script types.
+
+## Step 0: Design Philosophy
+
+Bitcoin-S strives to have script types defined in such a way that they can be easily composed and reused. Before going through this guide and implementing a really large script template type, try to decompose your script into smaller re-usable pieces.
+
+Also remember to consider what existing pieces you can use. For example, `LockTimeScriptPubKey`s are implemented in such a way that any other `RawScriptPubKey` can be given a time lock by nesting it within a `LockTimeScriptPubKey` subtype. Likewise, `ConditionalScriptPubKey`s are built to allow any other `RawScriptPubKey` type to populate both the `OP_IF/OP_NOTIF` case and the `OP_ELSE` case.
+
+## Step 1: Create a New ScriptPubKey Trait
+
+Go to `ScriptPubKey.scala` and add a new trait:
+
+```scala
```

Comment:
So one of the nice things about mdoc is that it gives you compiler errors if your documents every become out of date. For instance, if we were to change the `RawScriptPubKey` name, how would we know we need to update this doc? That's the problem that `mdoc` solves:

To use this, you need to use 
```scala mdoc:compile-only
//blah blah blah
```

now when you do `docs/run` it will compile the scala code and give you a compiler error if it isn't valid.